--- a/lib/includes/nghttp2/nghttp2ver.h.in
+++ b/lib/includes/nghttp2/nghttp2ver.h.in
@@ -29,7 +29,7 @@
  * @macro
  * Version number of the nghttp2 library release
  */
-#define NGHTTP2_VERSION "@PACKAGE_VERSION@"
+#define NGHTTP2_VERSION "1.59.0"
 
 /**
  * @macro
@@ -37,6 +37,6 @@
  * release. This is a 24 bit number with 8 bits for major number, 8 bits
  * for minor and 8 bits for patch. Version 1.2.3 becomes 0x010203.
  */
-#define NGHTTP2_VERSION_NUM @PACKAGE_VERSION_NUM@
+#define NGHTTP2_VERSION_NUM 0x013b00
 
 #endif /* NGHTTP2VER_H */
